apps/ledger-live-mobile/src/screens/SyncOnboarding/LanguageSelect.tsx

```diff
@@ -0,0 +1,213 @@
+
+type UiDrawerStatus =
+  | "none"
+  | "language-selection"
+  | "firmware-language-update";
+
+type LanguageSelectStatus =
+  | "unrequested"
+  | "language-selection-requested"
+  | "firmware-language-update-requested"
+  | "completed";
+
+export type Props = {
+  device: Device;
+  productName: string;
+};
+
+const ScrollViewContainer = styled(ScrollView)`
+  height: 100%;
+`;
+
+const LanguageSelect = ({ device, productName }: Props) => {
+  const { t } = useTranslation();
+  const { locale: currentLocale } = useLocale();
+  const dispatch = useDispatch();
+  const { availableLanguages: firmwareAvailableLanguages, loaded } =
+    useAvailableLanguagesForDevice(device);
+
+  const [currentDisplayedDrawer, setCurrentDisplayedDrawer] =
+    useState<UiDrawerStatus>("none");
+
+  // Will be computed depending on the states. Updating nextDrawerToDisplay
+  // triggers the current displayed drawer to close
+  let nextDrawerToDisplay: UiDrawerStatus = "none";
+
+  const [selectedLanguage, setSelectedLanguage] = useState<string | null>(null);
+
+  const [languageSelectStatus, setLanguageSelectStatus] =
+    useState<LanguageSelectStatus>("unrequested");
+
+  // Handles a newly selected language to redux-dispatch
+  useEffect(() => {
+    if (selectedLanguage) {
+      dispatch(setLanguage(selectedLanguage));
+    }
+  }, [dispatch, selectedLanguage]);
+
+  const handleLanguageSelectOnChange = useCallback(
+    language => {
+      setSelectedLanguage(language);
+
+      if (loaded && firmwareAvailableLanguages.includes(language)) {
+        setLanguageSelectStatus("firmware-language-update-requested");
+      } else {
+        setLanguageSelectStatus("completed");
+      }
+    },
+    [firmwareAvailableLanguages, loaded],
+  );
+
+  const handleLanguageSelectOnPress = useCallback(() => {
+    setLanguageSelectStatus("language-selection-requested");
+  }, []);
+
+  const handleLanguageSelectCancel = useCallback(() => {
+    setLanguageSelectStatus("completed");
+  }, []);
+
+  const handleFirmwareLanguageUpdate = useCallback(() => {
+    // TODO: redirect to firmware localization flow when available
+  }, []);
+
+  const handleFirmwareLanguageCancel = useCallback(() => {
+    setLanguageSelectStatus("completed");
+  }, []);
+
+  // Handles the UI logic
+  if (languageSelectStatus === "language-selection-requested") {
+    nextDrawerToDisplay = "language-selection";
+  } else if (languageSelectStatus === "firmware-language-update-requested") {
+    nextDrawerToDisplay = "firmware-language-update";
+  } else {
+    nextDrawerToDisplay = "none";
+
+    // Resets entirely the drawer mechanism
+    if (currentDisplayedDrawer !== "none") {
+      setCurrentDisplayedDrawer("none");
+    }
+  }
+
+  // If there is already a displayed drawer, the currentDisplayedDrawer would be
+  // synchronized with nextDrawerToDisplay during the displayed drawer onClose event.
+  // Otherwise, currentDisplayDrawer needs to be set to nextDrawerToDisplay manually
```

Comment:
![raw](https://user-images.githubusercontent.com/91890529/195059686-00e1672d-5bd1-4f6b-8a46-5be7a5d331a7.png)

Reply:
I know I know ... but it's because of the drawer bugs on ios .. you cannot hide a drawer A and display a drawer B at the same time .. a new rendering needs to happen in between - so I use the `onClose` event of the drawer A to display the drawer B. But there are cases where drawer A is not displayed, so we need to display drawer B outside of the `onClose` event (that does not exist if drawer A is not displayed)